engine/include/Engine/Intermediate/RenderSystem.hpp

```diff
@@ -0,0 +1,125 @@
+    std::vector<Camera*> m_pCameras;
+    std::vector<Light*>  m_pLights;
+
+    unsigned int               m_currentShaderId                  = 0;
+    unsigned int               m_currentTextureId                 = 0;
+    unsigned int               m_currentMeshId                    = 0;
+    Shader* m_currentPShaderUse                = nullptr;
+    bool                       m_currentBackFaceCullingModeEnable = false;
+
+protected:
+    void tryToBindShader(Shader* pShader);
+    void tryToBindTexture(unsigned int textureId);
+    void tryToBindMesh(unsigned int meshId);
+    void tryToSetAlphaEnabled(bool alphaEnabled);
+    void tryToSetBackFaceCulling(bool useBackFaceCulling);
+
+    void resetCurrentRenderPassKey();
+
+    inline RenderSystem() noexcept = default;
+    ~RenderSystem() noexcept       = default;
+
+public:
+    RenderSystem(const RenderSystem& other) noexcept = delete;
+    RenderSystem(RenderSystem&& other) noexcept      = delete;
+    RenderSystem& operator=(RenderSystem const& other) noexcept = delete;
+    RenderSystem& operator=(RenderSystem&& other) noexcept = delete;
+
+    void draw() noexcept;
+
+    /**
+     * @brief Add Model (using key word this) on the physic system. This object will be updated by the physic system
+     *
+     * @param pModel
+     */
+    // TODO: Remove this shit and create variadic templated system
+    void addModel(Model* pModel) noexcept;
+
+    void updateModelPointer(Model* newPointerModel,
+                            Model* exPointerModel) noexcept;
+
+    void removeModel(Model* pModel) noexcept;
+
+    // TODO: Remove this shit and create variadic templated system
+    void addCamera(Camera* pCamera) noexcept;
+
+    void updateCameraPointer(Camera* newPointerCamera,
+                             Camera* exPointerCamera) noexcept;
+
+    void removeCamera(Camera* pCamera) noexcept;
+
+    // TODO: Remove this shit and create variadic templated system
+    void addLight(Light* pLight) noexcept;
+
+    void updateLightPointer(Light* newPointerLight,
+                            Light* exPointerLight) noexcept;
+
+    void removeLight(Light* pLight) noexcept;
+
+    /**
+     * @brief This is the static method that controls the access to the singleton
+     * instance. On the first run, it creates a singleton object and places it
+     * into the static field. On subsequent runs, it returns the client existing
+     * object stored in the static field.
+     *
+     * @param value
+     * @return RenderSystem*
+     */
+    static RenderSystem* getInstance() noexcept
+    {
+        // double same if to avoid to lock mutex
+        if (unlikely(m_pInstance == nullptr))
+        {
+            std::unique_lock lock(m_mutex);
+            if (unlikely(m_pInstance == nullptr))
+            {
+                m_pInstance = new RenderSystem();
+            }
+            return m_pInstance;
+        }
+
+        std::shared_lock lock(m_mutex);
+        return m_pInstance;
+    }
```

Comment:
We could lazy-load the `RenderSystem`, by moving its instance into this method (which, by the way, I suggest we rename get()):

```suggestion
    static RenderSystem& getInstance() noexcept
    {
        // IMPORTANT: if this approach is adopted, lock m_mutex in RenderSystem's constructor
        // IMPORTANT: if this approach is adopted, remove m_pInstance from data members
        static RenderSystem renderSys;
        return renderSys;
    }
```